src/topology/locally_constant/basic.lean

```diff
@@ -0,0 +1,260 @@
+let ⟨U, hU, hx, H⟩ := hf.exists_open x in ⟨U, mem_nhds_sets hU hx, H⟩
+
+lemma iff_exists_open (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ (U : set X) (hU : is_open U) (hx : x ∈ U), ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_open, _⟩,
+  assume h s,
+  rw is_open_iff_forall_mem_open,
+  assume x hx,
+  obtain ⟨U, hU, hxU, H⟩ := h x,
+  refine ⟨U, _, hU, hxU⟩,
+  assume x' hx',
+  simp only [*, set.mem_preimage] at *,
+end
+
+lemma iff_exists_nhds (f : X → Y) :
+  is_locally_constant f ↔ ∀ x, ∃ U ∈ 𝓝 x, ∀ x' ∈ U, f x' = f x :=
+begin
+  refine ⟨exists_nhds, _⟩,
+  assume h,
+  rw iff_exists_open,
+  assume x,
+  obtain ⟨U, hU, H⟩ := h x,
+  obtain ⟨V, hVU, hV, hxV⟩ : ∃ (V : set X) (H : V ⊆ U), is_open V ∧ x ∈ V,
+  by rwa mem_nhds_sets_iff at hU,
+  refine ⟨V, hV, hxV, _⟩,
+  assume x' hx',
+  solve_by_elim only [H, hxV, hx', hVU]
+end
+
+lemma of_constant (f : X → Y) (h : ∃ y, ∀ x, f x = y) :
+  is_locally_constant f :=
+begin
+  obtain ⟨y, hy⟩ := h,
+  rw iff_exists_nhds,
+  intro x,
+  refine ⟨set.univ, filter.univ_mem_sets, _⟩,
+  rintro x -,
+  rw [hy, hy]
+end
+
+lemma const (y : Y) : is_locally_constant (function.const X y) :=
+of_constant _ ⟨y, λ _, rfl⟩
+
+protected lemma continuous {_ : topological_space Y} {f : X → Y} (hf : is_locally_constant f) :
+  continuous f :=
+⟨λ U hU, hf _⟩
+
+lemma iff_continuous {_ : topological_space Y} [discrete_topology Y] (f : X → Y) :
+  is_locally_constant f ↔ continuous f :=
+⟨is_locally_constant.continuous, λ h s, h.is_open_preimage s (is_open_discrete _)⟩
+
+lemma iff_continuous_bot (f : X → Y) :
+  is_locally_constant f ↔ @continuous X Y _ ⊥ f :=
+iff_continuous f
+
+lemma apply_eq_of_is_preconnected {f : X → Y} (hf : is_locally_constant f)
+  (s : set X) (hs : is_preconnected s) (x y : X) (hx : x ∈ s) (hy : y ∈ s) :
+  f y = f x :=
+begin
+  letI : topological_space Y := ⊥,
+  haveI : discrete_topology Y := ⟨rfl⟩,
+  have aux := is_preconnected.image hs f hf.continuous.continuous_on {f x} (f '' s \ {f x})
+    (is_open_discrete _) (is_open_discrete _),
+  simp only [set.union_diff_self, ← set.inter_diff_assoc, set.inter_self, set.inter_diff_self,
+    set.inter_empty, ← @set.ne_empty_iff_nonempty _ ∅, eq_self_iff_true, not_true, ne.def] at aux,
+  classical, by_contra hxy,
+  exact aux (set.subset_union_right _ _)
+    ⟨f x, set.mem_inter (set.mem_image_of_mem f hx) (set.mem_singleton _)⟩
+    ⟨f y, set.mem_diff_singleton.mpr ⟨set.mem_image_of_mem f hy, hxy⟩⟩
+end
+
+lemma range_finite [compact_space X] {f : X → Y} (hf : is_locally_constant f) :
+  (set.range f).finite :=
+begin
+  letI : topological_space Y := ⊥,
+  haveI : discrete_topology Y := ⟨rfl⟩,
+  rw @iff_continuous X Y ‹_› ‹_› at hf,
+  exact finite_of_is_compact_of_discrete _ (compact_range hf)
+end
+
+@[to_additive]
+lemma one [has_one Y] : is_locally_constant (1 : X → Y) := const 1
+
+@[to_additive]
+lemma inv [has_inv Y] ⦃f : X → Y⦄ (hf : is_locally_constant f) :
+  is_locally_constant f⁻¹ :=
```

Comment:
A more general lemma: if `f` is locally constant and `g` is any function, then `g ∘ f` is locally constant, and similarly with any binop.

Reply:
Fixed.